<!DOCTYPE HTML><html lang="zh-CN"><head><meta charset="UTF-8"><meta name="google-site-verification" content="GkJ68PCnxM9Ih7Zm9v8p91FOCV7ymNCO2KdGbflLh3E"><meta name="360-site-verification" content="dd6547587641bfede036f7c79561e8a0"><meta name="shenma-site-verification" content="fd96cf3751972c9b05f8471d2ccadddc_1496951214"><meta name="msvalidate.01" content="1D38B05050A977F6FDEDA481198343F1"><meta name="sogou_site_verification" content="MpPsku240L"><title>Hystrix 源码解析 —— 请求执行（四）之失败回退逻辑 | 芋道源码 —— 纯源码解析博客</title><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=3,minimum-scale=1"><meta name="author" content="芋道源码"><meta name="description" content="摘要: 原创出处 http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/ 「芋道源码」欢迎转载，保留摘要，谢谢！
本文主要基于 Hystrix 1.5.X 版本  

1. 概述
2. handleFallback
3. #han"><meta name="keywords" content="Java,架构,后端,服务端,RocketMQ,MyCAT,分布式消息队列,分布式存储,技术博客,Sharding-JDBC,分库分表"><link rel="alternate" href="atom.xml" title="芋道源码 —— 纯源码解析博客" type="application/atom+xml"><link rel="icon" href="/images/favicon.ico"><link rel="stylesheet" href="/css/style.css"><script>var _hmt=_hmt||[];!function(){var e=document.createElement("script");e.src="//hm.baidu.com/hm.js?9e70e3362807c1bd185a79655b307027";var t=document.getElementsByTagName("script")[0];t.parentNode.insertBefore(e,t)}()</script><script>!function(){var t=document.createElement("script"),e=window.location.protocol.split(":")[0];t.src="https"===e?"https://zz.bdstatic.com/linksubmit/push.js":"http://push.zhanzhang.baidu.com/push.js";var s=document.getElementsByTagName("script")[0];s.parentNode.insertBefore(t,s)}()</script><script>!function(){var t="http:"==document.location.protocol?"http://js.passport.qihucdn.com/11.0.1.js?f05b61dd54bbc38386611a5238672938":"https://jspassport.ssl.qhimg.com/11.0.1.js?f05b61dd54bbc38386611a5238672938";document.write('<script src="'+t+'" id="sozz"><\/script>')}()</script><script async src="//dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script></head></html><body><header><div><div id="textlogo"><h1 class="site-name"><a href="/" title="芋道源码 —— 纯源码解析博客">芋道源码 —— 纯源码解析博客</a></h1><a class="blog-motto">愿半生编码，如一生老友！</a></div><div class="navbar"><a class="navbutton navmobile" href="#" title="菜单"></a></div><nav class="animated"><ul><ul><li><a href="/">文章</a></li><li><a href="/2018-meet-you">知识星球</a></li><li><a href="https://github.com/YunaiV" rel="external nofollow noopener noreferrer" target="_blank">Github</a></li><li><a href="http://www.iocoder.cn/images/common/wechat_mp_2017_07_31_bak.jpg">微信公众号</a></li><li><a href="/categories/%E5%B7%A5%E4%BD%9C%E5%86%85%E6%8E%A8/">工作内推</a></li><li><a href="/link_url">友链</a></li></ul></ul></nav></div></header><div id="container"><div id="main" class="post" itemscope="" itemprop="blogPost"><article itemprop="articleBody"><header class="article-info clearfix"><h1 itemprop="name"><a href="/Hystrix/command-execute-fourth-fallback/" title="Hystrix 源码解析 —— 请求执行（四）之失败回退逻辑" itemprop="url">Hystrix 源码解析 —— 请求执行（四）之失败回退逻辑</a></h1><p class="article-author"></p><p class="article-time">总阅读量:<span id="busuanzi_value_page_pv"></span>次</p></header><div class="article-content"><p>摘要: 原创出处 <a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/</a> 「芋道源码」欢迎转载，保留摘要，谢谢！</p><p><strong>本文主要基于 Hystrix 1.5.X 版本</strong></p><ul><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">1. 概述</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">2. handleFallback</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">3. #handleShortCircuitViaFallback()</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">4. #handleSemaphoreRejectionViaFallback()</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">5. #handleThreadPoolRejectionViaFallback()</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">6. #handleTimeoutViaFallback()</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">7. #handleFailureViaFallback()</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">8. #getFallbackOrThrowException(…)</a></li><li><a href="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/">666. 彩蛋</a></li></ul><hr><p><img src="http://www.iocoder.cn/images/common/wechat_mp_2017_07_31.jpg" alt=""></p><blockquote><p>🙂🙂🙂关注<strong>微信公众号：【芋道源码】</strong>有福利：</p><ol><li>RocketMQ / MyCAT / Sharding-JDBC <strong>所有</strong>源码分析文章列表</li><li>RocketMQ / MyCAT / Sharding-JDBC <strong>中文注释源码 GitHub 地址</strong></li><li>您对于源码的疑问每条留言<strong>都</strong>将得到<strong>认真</strong>回复。<strong>甚至不知道如何读源码也可以请教噢</strong>。</li><li><strong>新的</strong>源码解析文章<strong>实时</strong>收到通知。<strong>每周更新一篇左右</strong>。</li><li><strong>认真的</strong>源码交流微信群。</li></ol></blockquote><hr><h1 id="1-概述"><a href="#1-概述" class="headerlink" title="1. 概述"></a>1. 概述</h1><p>本文主要分享 <strong>Hystrix 命令执行（四）之失败回退逻辑</strong>。</p><p>建议 ：对 RxJava 已经有一定的了解的基础上阅读本文。</p><p>Hystrix 执行命令整体流程如下图：</p><blockquote><p>FROM <a href="http://youdang.github.io/2016/02/05/translate-hystrix-wiki-how-it-works/#流程图" rel="external nofollow noopener noreferrer" target="_blank">《【翻译】Hystrix文档-实现原理》「流程图」</a><br><img src="http://www.iocoder.cn/images/Hystrix/2018_10_31/01.jpeg" alt=""></p></blockquote><ul><li><strong>红</strong>圈 ：Hystrix 命令执行失败，执行回退逻辑。<strong>也就是大家经常在文章中看到的“服务降级”</strong>。</li><li><strong>绿</strong>圈 ：四种情况会触发失败回退逻辑( fallback )。<ul><li>第一种 ：<code>short-circuit</code> ，处理<strong>链路处于熔断</strong>的回退逻辑，在 <a href="#">「3. #handleShortCircuitViaFallback()」</a> 详细解析。</li><li>第二种 ：<code>semaphore-rejection</code> ，处理<strong>信号量获得失败</strong>的回退逻辑，在 <a href="#">「4. #handleShortCircuitViaFallback()」</a> 详细解析。</li><li>第三种 ：<code>thread-pool-rejection</code> ，处理<strong>线程池提交任务拒绝</strong>的回退逻辑，在 <a href="#">「5. #handleThreadPoolRejectionViaFallback()」</a> 详细解析。</li><li>第四种 ：<code>execution-timeout</code> ，处理<strong>命令执行超时</strong>的回退逻辑，在 <a href="#">「6. #handleTimeoutViaFallback()」</a> 详细解析。</li><li>第五种 ：<code>execution-failure</code> ，处理<strong>命令执行异常</strong>的回退逻辑，在 <a href="#">「7. #handleFailureViaFallback()」</a> 详细解析。</li><li>第六种 ：<code>bad-request</code> ，TODO 【2014】【HystrixBadRequestException】，和 <code>hystrix-javanica</code> 子项目相关。</li></ul></li></ul><p>另外，<code>#handleXXXX()</code> 方法，整体代码比较类似，最终都是调用 <code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，在 <a href="#">「8. #getFallbackOrThrowException(…)」</a> 详细解析。</p><hr><p><strong>推荐 Spring Cloud 书籍</strong>：</p><ul><li>请支持正版。下载盗版，<strong>等于主动编写低级 BUG</strong> 。</li><li>程序猿DD —— <a href="https://union-click.jd.com/jdc?d=505Twi" rel="external nofollow noopener noreferrer" target="_blank">《Spring Cloud微服务实战》</a></li><li>周立 —— <a href="https://union-click.jd.com/jdc?d=k3sAaK" rel="external nofollow noopener noreferrer" target="_blank">《Spring Cloud与Docker微服务架构实战》</a></li><li>两书齐买，京东包邮。</li></ul><p><strong>推荐 Spring Cloud 视频</strong>：</p><ul><li><a href="https://segmentfault.com/ls/1650000011063780?r=bPN0Ir" rel="external nofollow noopener noreferrer" target="_blank">Java 微服务实践 - Spring Boot</a></li><li><a href="https://segmentfault.com/ls/1650000011386794?r=bPN0Ir" rel="external nofollow noopener noreferrer" target="_blank">Java 微服务实践 - Spring Cloud</a></li><li><a href="https://segmentfault.com/ls/1650000011387052?r=bPN0Ir" rel="external nofollow noopener noreferrer" target="_blank">Java 微服务实践 - Spring Boot / Spring Cloud</a></li></ul><h1 id="2-handleFallback"><a href="#2-handleFallback" class="headerlink" title="2. handleFallback"></a>2. handleFallback</h1><p>在 <a href="http://www.iocoder.cn/Hystrix/command-execute-first-run/?self">《Hystrix 源码解析 —— 命令执行（一）之正常执行逻辑》「4. #executeCommandAndObserve(…)」</a> 中，<code>#executeCommandAndObserve(...)</code> 的<strong>第 82 行</strong> <code>onErrorResumeNext(handleFallback)</code> 代码，通过调用 <code>Observable#onErrorResumeNext(...)</code> 方法，实现【执行命令 Observable】执行异常时，返回【回退逻辑 Observable】，执行失败回退逻辑。</p><blockquote><p>FROM <a href="https://mcxiaoke.gitbooks.io/rxdocs/content/operators/Catch.html#onerrorreturn" rel="external nofollow noopener noreferrer" target="_blank">《ReactiveX文档中文翻译》「onErrorResumeNext」</a><br>onErrorResumeNext 方法返回一个镜像原有 Observable 行为的新 Observable ，后者会忽略前者的 onError 调用，不会将错误传递给观察者，作为替代，它会开始镜像另一个，备用的 Observable 。</p><ul><li>Javadoc: <a href="http://reactivex.io/RxJava/javadoc/rx/Observable.html#onErrorResumeNext-rx.functions.Func1-" rel="external nofollow noopener noreferrer" target="_blank">onErrorResumeNext(Func1))</a></li><li>Javadoc: <a href="http://reactivex.io/RxJava/javadoc/rx/Observable.html#onErrorResumeNext-rx.Observable-" rel="external nofollow noopener noreferrer" target="_blank">onErrorResumeNext(Observable))</a></li></ul></blockquote><p><a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L602" rel="external nofollow noopener noreferrer" target="_blank"><code>handleFallback</code></a> 变量，代码如下 ：</p><figure class="highlight java"><table><tr><td class="code"><pre><div class="line"> <span class="number">1</span>: <span class="keyword">final</span> Func1&lt;Throwable, Observable&lt;R&gt;&gt; handleFallback = <span class="keyword">new</span> Func1&lt;Throwable, Observable&lt;R&gt;&gt;() &#123;</div><div class="line"> <span class="number">2</span>:     <span class="meta">@Override</span></div><div class="line"> <span class="number">3</span>:     <span class="function"><span class="keyword">public</span> Observable&lt;R&gt; <span class="title">call</span><span class="params">(Throwable t)</span> </span>&#123;</div><div class="line"> <span class="number">4</span>:         <span class="comment">// 标记尝试成功</span></div><div class="line"> <span class="number">5</span>:         circuitBreaker.markNonSuccess();</div><div class="line"> <span class="number">6</span>:         <span class="comment">// 标记 executionResult 执行异常</span></div><div class="line"> <span class="number">7</span>:         Exception e = getExceptionFromThrowable(t);</div><div class="line"> <span class="number">8</span>:         executionResult = executionResult.setExecutionException(e);</div><div class="line"> <span class="number">9</span>:         <span class="comment">// 返回 【回退逻辑 Observable】</span></div><div class="line"><span class="number">10</span>:         <span class="keyword">if</span> (e <span class="keyword">instanceof</span> RejectedExecutionException) &#123; <span class="comment">// 线程池提交任务拒绝异常</span></div><div class="line"><span class="number">11</span>:             <span class="keyword">return</span> handleThreadPoolRejectionViaFallback(e);</div><div class="line"><span class="number">12</span>:         &#125; <span class="keyword">else</span> <span class="keyword">if</span> (t <span class="keyword">instanceof</span> HystrixTimeoutException) &#123; <span class="comment">// 执行命令超时异常</span></div><div class="line"><span class="number">13</span>:             <span class="keyword">return</span> handleTimeoutViaFallback();</div><div class="line"><span class="number">14</span>:         &#125; <span class="keyword">else</span> <span class="keyword">if</span> (t <span class="keyword">instanceof</span> HystrixBadRequestException) &#123; <span class="comment">// TODO 【2014】【HystrixBadRequestException】</span></div><div class="line"><span class="number">15</span>:             <span class="keyword">return</span> handleBadRequestByEmittingError(e);</div><div class="line"><span class="number">16</span>:         &#125; <span class="keyword">else</span> &#123;</div><div class="line"><span class="number">17</span>:             <span class="comment">/*</span></div><div class="line"><span class="comment">18:              * Treat HystrixBadRequestException from ExecutionHook like a plain HystrixBadRequestException.</span></div><div class="line"><span class="comment">19:              */</span></div><div class="line"><span class="number">20</span>:             <span class="keyword">if</span> (e <span class="keyword">instanceof</span> HystrixBadRequestException) &#123; <span class="comment">// TODO 【2014】【HystrixBadRequestException】</span></div><div class="line"><span class="number">21</span>:                 eventNotifier.markEvent(HystrixEventType.BAD_REQUEST, commandKey);</div><div class="line"><span class="number">22</span>:                 <span class="keyword">return</span> Observable.error(e);</div><div class="line"><span class="number">23</span>:             &#125;</div><div class="line"><span class="number">24</span>: </div><div class="line"><span class="number">25</span>:             <span class="keyword">return</span> handleFailureViaFallback(e);</div><div class="line"><span class="number">26</span>:         &#125;</div><div class="line"><span class="number">27</span>:     &#125;</div><div class="line"><span class="number">28</span>: &#125;;</div></pre></td></tr></table></figure><ul><li>第 5 行 ：标记断路器尝试成功。在 <a href="http://www.iocoder.cn/Hystrix/circuit-breaker/?self">《Hystrix 源码解析 —— 断路器 HystrixCircuitBreaker》</a> 有详细解析。</li><li>第 7 至 8 行 ：标记 <code>executionResult</code> <strong>执行异常</strong>。</li><li>第 10 至 11 行 ：<code>thread-pool-rejection</code> ，处理<strong>线程池提交任务拒绝</strong>的回退逻辑，在 <a href="#">「5. #handleThreadPoolRejectionViaFallback()」</a> 详细解析。</li><li>第 12 至 13 行 ：<code>execution-timeout</code> ，处理<strong>命令执行超时</strong>的回退逻辑，在 <a href="">「6. #handleTimeoutViaFallback()」</a> 详细解析。</li><li>第 14 至 23 行 ：，<code>bad-request</code> ，TODO 【2014】【HystrixBadRequestException】，和 <code>hystrix-javanica</code> 子项目相关。</li><li>第 25 行 ：<code>execution-failure</code> 处理<strong>命令执行异常</strong>的回退逻辑，在 <a href="#">「7. #handleFailureViaFallback()」</a> 详细解析。</li></ul><h1 id="3-handleShortCircuitViaFallback"><a href="#3-handleShortCircuitViaFallback" class="headerlink" title="3. #handleShortCircuitViaFallback()"></a>3. #handleShortCircuitViaFallback()</h1><p><code>#handleShortCircuitViaFallback()</code> 方法，<code>short-circuit</code> ，处理<strong>链路处于熔断</strong>的回退逻辑，在 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L558" rel="external nofollow noopener noreferrer" target="_blank">此处</a> 被调用，代码如下 ：</p><figure class="highlight plain"><table><tr><td class="code"><pre><div class="line"> 1: private Observable&lt;R&gt; handleShortCircuitViaFallback() &#123;</div><div class="line"> 2:     // TODO 【2011】【Hystrix 事件机制】</div><div class="line"> 3:     // record that we are returning a short-circuited fallback</div><div class="line"> 4:     eventNotifier.markEvent(HystrixEventType.SHORT_CIRCUITED, commandKey);</div><div class="line"> 5:     // 标记 executionResult 执行异常</div><div class="line"> 6:     // short-circuit and go directly to fallback (or throw an exception if no fallback implemented)</div><div class="line"> 7:     Exception shortCircuitException = new RuntimeException(&quot;Hystrix circuit short-circuited and is OPEN&quot;);</div><div class="line"> 8:     executionResult = executionResult.setExecutionException(shortCircuitException);</div><div class="line"> 9:     try &#123;</div><div class="line">10:         // 获得 【回退逻辑 Observable】 或者 【异常 Observable】</div><div class="line">11:         return getFallbackOrThrowException(this, HystrixEventType.SHORT_CIRCUITED, FailureType.SHORTCIRCUIT,</div><div class="line">12:                 &quot;short-circuited&quot;, shortCircuitException);</div><div class="line">13:     &#125; catch (Exception e) &#123;</div><div class="line">14:         return Observable.error(e);</div><div class="line">15:     &#125;</div><div class="line">16: &#125;</div></pre></td></tr></table></figure><ul><li>第 4 行 ：TODO 【2011】【Hystrix 事件机制】</li><li>第 7 至 8 行 ：标记 <code>executionResult</code> <strong>执行异常</strong>。</li><li>第 11 至 12 行 ：调用 <code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，在 <a href="#">「8. #getFallbackOrThrowException(…)」</a> 详细解析。</li><li>第 14 行 ：返回【异常 Observable】。</li></ul><h1 id="4-handleSemaphoreRejectionViaFallback"><a href="#4-handleSemaphoreRejectionViaFallback" class="headerlink" title="4. #handleSemaphoreRejectionViaFallback()"></a>4. #handleSemaphoreRejectionViaFallback()</h1><p><code>#handleSemaphoreRejectionViaFallback()</code> 方法，<code>semaphore-rejection</code> ，处理<strong>信号量获得失败</strong>的回退逻辑，在 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L555" rel="external nofollow noopener noreferrer" target="_blank">此处</a> 被调用，代码如下 ：</p><figure class="highlight java"><table><tr><td class="code"><pre><div class="line"> <span class="number">1</span>: <span class="function"><span class="keyword">private</span> Observable&lt;R&gt; <span class="title">handleSemaphoreRejectionViaFallback</span><span class="params">()</span> </span>&#123;</div><div class="line"> <span class="number">2</span>:     <span class="comment">// 标记 executionResult 执行异常</span></div><div class="line"> <span class="number">3</span>:     Exception semaphoreRejectionException = <span class="keyword">new</span> RuntimeException(<span class="string">"could not acquire a semaphore for execution"</span>);</div><div class="line"> <span class="number">4</span>:     executionResult = executionResult.setExecutionException(semaphoreRejectionException);</div><div class="line"> <span class="number">5</span>:     <span class="comment">// TODO 【2011】【Hystrix 事件机制】</span></div><div class="line"> <span class="number">6</span>:     eventNotifier.markEvent(HystrixEventType.SEMAPHORE_REJECTED, commandKey);</div><div class="line"> <span class="number">7</span>:     logger.debug(<span class="string">"HystrixCommand Execution Rejection by Semaphore."</span>); <span class="comment">// debug only since we're throwing the exception and someone higher will do something with it</span></div><div class="line"> <span class="number">8</span>:     <span class="comment">// retrieve a fallback or throw an exception if no fallback available</span></div><div class="line"> <span class="number">9</span>:     <span class="comment">// 获得 【回退逻辑 Observable】 或者 【异常 Observable】</span></div><div class="line"><span class="number">10</span>:     <span class="keyword">return</span> getFallbackOrThrowException(<span class="keyword">this</span>, HystrixEventType.SEMAPHORE_REJECTED, FailureType.REJECTED_SEMAPHORE_EXECUTION,</div><div class="line"><span class="number">11</span>:             <span class="string">"could not acquire a semaphore for execution"</span>, semaphoreRejectionException);</div><div class="line"><span class="number">12</span>: &#125;</div></pre></td></tr></table></figure><ul><li>第 3 至 4 行 ：标记 <code>executionResult</code> <strong>执行异常</strong>。</li><li>第 6 至 7 行 ：TODO 【2011】【Hystrix 事件机制】</li><li>第 10 至 11 行 ：调用 <code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，在 <a href="#">「8. #getFallbackOrThrowException(…)」</a> 详细解析。</li></ul><h1 id="5-handleThreadPoolRejectionViaFallback"><a href="#5-handleThreadPoolRejectionViaFallback" class="headerlink" title="5. #handleThreadPoolRejectionViaFallback()"></a>5. #handleThreadPoolRejectionViaFallback()</h1><p><code>#handleThreadPoolRejectionViaFallback()</code> 方法，<code>thread-pool-rejection</code> ，处理<strong>线程池提交任务拒绝</strong>的回退逻辑，在 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L990" rel="external nofollow noopener noreferrer" target="_blank">此处</a> 被调用，代码如下：</p><figure class="highlight java"><table><tr><td class="code"><pre><div class="line"> <span class="number">1</span>: <span class="function"><span class="keyword">private</span> Observable&lt;R&gt; <span class="title">handleThreadPoolRejectionViaFallback</span><span class="params">(Exception underlying)</span> </span>&#123;</div><div class="line"> <span class="number">2</span>:     <span class="comment">// TODO 【2011】【Hystrix 事件机制】</span></div><div class="line"> <span class="number">3</span>:     eventNotifier.markEvent(HystrixEventType.THREAD_POOL_REJECTED, commandKey);</div><div class="line"> <span class="number">4</span>:     <span class="comment">// TODO 【2002】【metrics】</span></div><div class="line"> <span class="number">5</span>:     threadPool.markThreadRejection();</div><div class="line"> <span class="number">6</span>:     <span class="comment">// 获得 【回退逻辑 Observable】 或者 【异常 Observable】</span></div><div class="line"> <span class="number">7</span>:     <span class="comment">// use a fallback instead (or throw exception if not implemented)</span></div><div class="line"> <span class="number">8</span>:     <span class="keyword">return</span> getFallbackOrThrowException(<span class="keyword">this</span>, HystrixEventType.THREAD_POOL_REJECTED, FailureType.REJECTED_THREAD_EXECUTION,</div><div class="line"> <span class="number">9</span>:             <span class="string">"could not be queued for execution"</span>, underlying);</div><div class="line"><span class="number">10</span>: &#125;</div></pre></td></tr></table></figure><ul><li>第 3 行 ：TODO 【2011】【Hystrix 事件机制】</li><li>第 5 行 ：TODO 【2002】【metrics】</li><li>第 8 至 9 行 ：调用 <code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，在 <a href="#">「8. #getFallbackOrThrowException(…)」</a> 详细解析。</li></ul><h1 id="6-handleTimeoutViaFallback"><a href="#6-handleTimeoutViaFallback" class="headerlink" title="6. #handleTimeoutViaFallback()"></a>6. #handleTimeoutViaFallback()</h1><p><code>#handleTimeoutViaFallback()</code> 方法，<code>execution-timeout</code> ，处理<strong>命令执行超时</strong>的回退逻辑，在 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L611" rel="external nofollow noopener noreferrer" target="_blank">此处</a> 被调用，代码如下：</p><figure class="highlight java"><table><tr><td class="code"><pre><div class="line"><span class="number">1</span>: <span class="function"><span class="keyword">private</span> Observable&lt;R&gt; <span class="title">handleTimeoutViaFallback</span><span class="params">()</span> </span>&#123;</div><div class="line"><span class="number">2</span>:     <span class="comment">// 获得 【回退逻辑 Observable】 或者 【异常 Observable】</span></div><div class="line"><span class="number">3</span>:     <span class="keyword">return</span> getFallbackOrThrowException(<span class="keyword">this</span>, HystrixEventType.TIMEOUT, FailureType.TIMEOUT,</div><div class="line"><span class="number">4</span>:             <span class="string">"timed-out"</span>, <span class="keyword">new</span> TimeoutException());</div><div class="line"><span class="number">5</span>: &#125;</div></pre></td></tr></table></figure><ul><li>第 3 至 4 行 ：调用 <code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，在 <a href="#">「8. #getFallbackOrThrowException(…)」</a> 详细解析。</li></ul><h1 id="7-handleFailureViaFallback"><a href="#7-handleFailureViaFallback" class="headerlink" title="7. #handleFailureViaFallback()"></a>7. #handleFailureViaFallback()</h1><p><code>#handleFailureViaFallback()</code> 方法，<code>execution-failure</code> ，处理<strong>命令执行异常</strong>的回退逻辑，在 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L623" rel="external nofollow noopener noreferrer" target="_blank">此处</a> 被调用，代码如下：</p><figure class="highlight java"><table><tr><td class="code"><pre><div class="line"> <span class="number">1</span>: <span class="function"><span class="keyword">private</span> Observable&lt;R&gt; <span class="title">handleFailureViaFallback</span><span class="params">(Exception underlying)</span> </span>&#123;</div><div class="line"> <span class="number">2</span>:     <span class="comment">// TODO 【2011】【Hystrix 事件机制】</span></div><div class="line"> <span class="number">3</span>:     <span class="comment">/**</span></div><div class="line"><span class="comment"> 4:      * All other error handling</span></div><div class="line"><span class="comment"> 5:      */</span></div><div class="line"> <span class="number">6</span>:     logger.debug(<span class="string">"Error executing HystrixCommand.run(). Proceeding to fallback logic ..."</span>, underlying);</div><div class="line"> <span class="number">7</span>: </div><div class="line"> <span class="number">8</span>:     <span class="comment">// report failure</span></div><div class="line"> <span class="number">9</span>:     eventNotifier.markEvent(HystrixEventType.FAILURE, commandKey);</div><div class="line"><span class="number">10</span>: </div><div class="line"><span class="number">11</span>:     <span class="comment">// 标记 executionResult 异常 TODO 【2007】【executionResult】用途 为啥不是执行异常</span></div><div class="line"><span class="number">12</span>:     <span class="comment">// record the exception</span></div><div class="line"><span class="number">13</span>:     executionResult = executionResult.setException(underlying);</div><div class="line"><span class="number">14</span>:     <span class="comment">// 获得 【回退逻辑 Observable】 或者 【异常 Observable】</span></div><div class="line"><span class="number">15</span>:     <span class="keyword">return</span> getFallbackOrThrowException(<span class="keyword">this</span>, HystrixEventType.FAILURE, FailureType.COMMAND_EXCEPTION, <span class="string">"failed"</span>, underlying);</div><div class="line"><span class="number">16</span>: &#125;</div></pre></td></tr></table></figure><ul><li>第 2 至 9 行 ：TODO 【2011】【Hystrix 事件机制】</li><li>第 13 行 ：标记 <code>executionResult</code> <strong>异常</strong>。</li><li>第 15 行 ：调用 <code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，在 <a href="#">「8. #getFallbackOrThrowException(…)」</a> 详细解析。</li></ul><h1 id="8-getFallbackOrThrowException-…"><a href="#8-getFallbackOrThrowException-…" class="headerlink" title="8. #getFallbackOrThrowException(…)"></a>8. #getFallbackOrThrowException(…)</h1><p><code>#getFallbackOrThrowException()</code> 方法，获得【回退逻辑 Observable】或者【异常 Observable】，代码如下 ：</p><figure class="highlight java"><table><tr><td class="code"><pre><div class="line">  <span class="number">1</span>: <span class="function"><span class="keyword">private</span> Observable&lt;R&gt; <span class="title">getFallbackOrThrowException</span><span class="params">(<span class="keyword">final</span> AbstractCommand&lt;R&gt; _cmd, <span class="keyword">final</span> HystrixEventType eventType, <span class="keyword">final</span> FailureType failureType, <span class="keyword">final</span> String message, <span class="keyword">final</span> Exception originalException)</span> </span>&#123;</div><div class="line">  <span class="number">2</span>:     <span class="comment">// 记录 HystrixRequestContext</span></div><div class="line">  <span class="number">3</span>:     <span class="keyword">final</span> HystrixRequestContext requestContext = HystrixRequestContext.getContextForCurrentThread();</div><div class="line">  <span class="number">4</span>:     <span class="comment">// 标记 executionResult 添加( 记录 )事件</span></div><div class="line">  <span class="number">5</span>:     <span class="keyword">long</span> latency = System.currentTimeMillis() - executionResult.getStartTimestamp();</div><div class="line">  <span class="number">6</span>:     <span class="comment">// record the executionResult</span></div><div class="line">  <span class="number">7</span>:     <span class="comment">// do this before executing fallback so it can be queried from within getFallback (see See https://github.com/Netflix/Hystrix/pull/144)</span></div><div class="line">  <span class="number">8</span>:     executionResult = executionResult.addEvent((<span class="keyword">int</span>) latency, eventType);</div><div class="line">  <span class="number">9</span>: </div><div class="line"> <span class="number">10</span>:     <span class="keyword">if</span> (isUnrecoverable(originalException)) &#123; <span class="comment">// 无法恢复的异常</span></div><div class="line"> <span class="number">11</span>:         logger.error(<span class="string">"Unrecoverable Error for HystrixCommand so will throw HystrixRuntimeException and not apply fallback. "</span>, originalException);</div><div class="line"> <span class="number">12</span>: </div><div class="line"> <span class="number">13</span>:         <span class="comment">// TODO 【2003】【HOOK】</span></div><div class="line"> <span class="number">14</span>:         <span class="comment">/* executionHook for all errors */</span></div><div class="line"> <span class="number">15</span>:         Exception e = wrapWithOnErrorHook(failureType, originalException);</div><div class="line"> <span class="number">16</span>:         <span class="comment">// 返回 【异常 Observable】</span></div><div class="line"> <span class="number">17</span>:         <span class="keyword">return</span> Observable.error(<span class="keyword">new</span> HystrixRuntimeException(failureType, <span class="keyword">this</span>.getClass(), getLogMessagePrefix() + <span class="string">" "</span> + message + <span class="string">" and encountered unrecoverable error."</span>, e, <span class="keyword">null</span>));</div><div class="line"> <span class="number">18</span>:     &#125; <span class="keyword">else</span> &#123;</div><div class="line"> <span class="number">19</span>:         <span class="keyword">if</span> (isRecoverableError(originalException)) &#123; <span class="comment">// 可恢复的异常</span></div><div class="line"> <span class="number">20</span>:             logger.warn(<span class="string">"Recovered from java.lang.Error by serving Hystrix fallback"</span>, originalException);</div><div class="line"> <span class="number">21</span>:         &#125;</div><div class="line"> <span class="number">22</span>: </div><div class="line"> <span class="number">23</span>:         <span class="keyword">if</span> (properties.fallbackEnabled().get()) &#123;</div><div class="line"> <span class="number">24</span>:             <span class="comment">/* fallback behavior is permitted so attempt */</span></div><div class="line"> <span class="number">25</span>: </div><div class="line"> <span class="number">26</span>:             <span class="comment">// 设置 HystrixRequestContext 的 Action</span></div><div class="line"> <span class="number">27</span>:             <span class="keyword">final</span> Action1&lt;Notification&lt;? <span class="keyword">super</span> R&gt;&gt; setRequestContext = <span class="keyword">new</span> Action1&lt;Notification&lt;? <span class="keyword">super</span> R&gt;&gt;() &#123;</div><div class="line"> <span class="number">28</span>:                 <span class="meta">@Override</span></div><div class="line"> <span class="number">29</span>:                 <span class="function"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title">call</span><span class="params">(Notification&lt;? <span class="keyword">super</span> R&gt; rNotification)</span> </span>&#123;</div><div class="line"> <span class="number">30</span>:                     setRequestContextIfNeeded(requestContext);</div><div class="line"> <span class="number">31</span>:                 &#125;</div><div class="line"> <span class="number">32</span>:             &#125;;</div><div class="line"> <span class="number">33</span>: </div><div class="line"> <span class="number">34</span>:             <span class="comment">// TODO 【2007】【executionResult】用途</span></div><div class="line"> <span class="number">35</span>:             <span class="keyword">final</span> Action1&lt;R&gt; markFallbackEmit = <span class="keyword">new</span> Action1&lt;R&gt;() &#123;</div><div class="line"> <span class="number">36</span>:                 <span class="meta">@Override</span></div><div class="line"> <span class="number">37</span>:                 <span class="function"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title">call</span><span class="params">(R r)</span> </span>&#123;</div><div class="line"> <span class="number">38</span>:                     <span class="keyword">if</span> (shouldOutputOnNextEvents()) &#123;</div><div class="line"> <span class="number">39</span>:                         executionResult = executionResult.addEvent(HystrixEventType.FALLBACK_EMIT);</div><div class="line"> <span class="number">40</span>:                         eventNotifier.markEvent(HystrixEventType.FALLBACK_EMIT, commandKey);</div><div class="line"> <span class="number">41</span>:                     &#125;</div><div class="line"> <span class="number">42</span>:                 &#125;</div><div class="line"> <span class="number">43</span>:             &#125;;</div><div class="line"> <span class="number">44</span>: </div><div class="line"> <span class="number">45</span>:             <span class="comment">// TODO 【2007】【executionResult】用途</span></div><div class="line"> <span class="number">46</span>:             <span class="keyword">final</span> Action0 markFallbackCompleted = <span class="keyword">new</span> Action0() &#123;</div><div class="line"> <span class="number">47</span>:                 <span class="meta">@Override</span></div><div class="line"> <span class="number">48</span>:                 <span class="function"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title">call</span><span class="params">()</span> </span>&#123;</div><div class="line"> <span class="number">49</span>:                     <span class="keyword">long</span> latency = System.currentTimeMillis() - executionResult.getStartTimestamp();</div><div class="line"> <span class="number">50</span>:                     eventNotifier.markEvent(HystrixEventType.FALLBACK_SUCCESS, commandKey);</div><div class="line"> <span class="number">51</span>:                     executionResult = executionResult.addEvent((<span class="keyword">int</span>) latency, HystrixEventType.FALLBACK_SUCCESS);</div><div class="line"> <span class="number">52</span>:                 &#125;</div><div class="line"> <span class="number">53</span>:             &#125;;</div><div class="line"> <span class="number">54</span>: </div><div class="line"> <span class="number">55</span>:             <span class="comment">// 处理异常 的 Func</span></div><div class="line"> <span class="number">56</span>:             <span class="keyword">final</span> Func1&lt;Throwable, Observable&lt;R&gt;&gt; handleFallbackError = <span class="keyword">new</span> Func1&lt;Throwable, Observable&lt;R&gt;&gt;() &#123;</div><div class="line"> <span class="number">57</span>:                 <span class="meta">@Override</span></div><div class="line"> <span class="number">58</span>:                 <span class="function"><span class="keyword">public</span> Observable&lt;R&gt; <span class="title">call</span><span class="params">(Throwable t)</span> </span>&#123;</div><div class="line"> <span class="number">59</span>:                     <span class="comment">// TODO 【2003】【HOOK】</span></div><div class="line"> <span class="number">60</span>:                     <span class="comment">/* executionHook for all errors */</span></div><div class="line"> <span class="number">61</span>:                     Exception e = wrapWithOnErrorHook(failureType, originalException);</div><div class="line"> <span class="number">62</span>:                     <span class="comment">// 获得 Exception</span></div><div class="line"> <span class="number">63</span>:                     Exception fe = getExceptionFromThrowable(t);</div><div class="line"> <span class="number">64</span>: </div><div class="line"> <span class="number">65</span>:                     <span class="keyword">long</span> latency = System.currentTimeMillis() - executionResult.getStartTimestamp();</div><div class="line"> <span class="number">66</span>:                     Exception toEmit;</div><div class="line"> <span class="number">67</span>: </div><div class="line"> <span class="number">68</span>:                     <span class="keyword">if</span> (fe <span class="keyword">instanceof</span> UnsupportedOperationException) &#123;</div><div class="line"> <span class="number">69</span>:                         <span class="comment">// TODO 【2011】【Hystrix 事件机制】</span></div><div class="line"> <span class="number">70</span>:                         logger.debug(<span class="string">"No fallback for HystrixCommand. "</span>, fe); <span class="comment">// debug only since we're throwing the exception and someone higher will do something with it</span></div><div class="line"> <span class="number">71</span>:                         eventNotifier.markEvent(HystrixEventType.FALLBACK_MISSING, commandKey);</div><div class="line"> <span class="number">72</span>:                         <span class="comment">// 标记 executionResult 添加( 记录 )事件 HystrixEventType.FALLBACK_MISSING</span></div><div class="line"> <span class="number">73</span>:                         executionResult = executionResult.addEvent((<span class="keyword">int</span>) latency, HystrixEventType.FALLBACK_MISSING);</div><div class="line"> <span class="number">74</span>: </div><div class="line"> <span class="number">75</span>:                         <span class="comment">// 创建 HystrixRuntimeException</span></div><div class="line"> <span class="number">76</span>:                         toEmit = <span class="keyword">new</span> HystrixRuntimeException(failureType, _cmd.getClass(), getLogMessagePrefix() + <span class="string">" "</span> + message + <span class="string">" and no fallback available."</span>, e, fe);</div><div class="line"> <span class="number">77</span>:                     &#125; <span class="keyword">else</span> &#123;</div><div class="line"> <span class="number">78</span>:                         <span class="comment">// TODO 【2011】【Hystrix 事件机制】</span></div><div class="line"> <span class="number">79</span>:                         logger.debug(<span class="string">"HystrixCommand execution "</span> + failureType.name() + <span class="string">" and fallback failed."</span>, fe);</div><div class="line"> <span class="number">80</span>:                         eventNotifier.markEvent(HystrixEventType.FALLBACK_FAILURE, commandKey);</div><div class="line"> <span class="number">81</span>:                         <span class="comment">// 标记 executionResult 添加( 记录 )事件 HystrixEventType.FALLBACK_FAILURE</span></div><div class="line"> <span class="number">82</span>:                         executionResult = executionResult.addEvent((<span class="keyword">int</span>) latency, HystrixEventType.FALLBACK_FAILURE);</div><div class="line"> <span class="number">83</span>: </div><div class="line"> <span class="number">84</span>:                         <span class="comment">// 创建 HystrixRuntimeException</span></div><div class="line"> <span class="number">85</span>:                         toEmit = <span class="keyword">new</span> HystrixRuntimeException(failureType, _cmd.getClass(), getLogMessagePrefix() + <span class="string">" "</span> + message + <span class="string">" and fallback failed."</span>, e, fe);</div><div class="line"> <span class="number">86</span>:                     &#125;</div><div class="line"> <span class="number">87</span>: </div><div class="line"> <span class="number">88</span>:                     <span class="comment">// <span class="doctag">NOTE:</span> we're suppressing fallback exception here</span></div><div class="line"> <span class="number">89</span>:                     <span class="keyword">if</span> (shouldNotBeWrapped(originalException)) &#123;</div><div class="line"> <span class="number">90</span>:                         <span class="keyword">return</span> Observable.error(e);</div><div class="line"> <span class="number">91</span>:                     &#125;</div><div class="line"> <span class="number">92</span>: </div><div class="line"> <span class="number">93</span>:                     <span class="keyword">return</span> Observable.error(toEmit);</div><div class="line"> <span class="number">94</span>:                 &#125;</div><div class="line"> <span class="number">95</span>:             &#125;;</div><div class="line"> <span class="number">96</span>: </div><div class="line"> <span class="number">97</span>:             <span class="comment">// 获得 TryableSemaphore</span></div><div class="line"> <span class="number">98</span>:             <span class="keyword">final</span> TryableSemaphore fallbackSemaphore = getFallbackSemaphore();</div><div class="line"> <span class="number">99</span>: </div><div class="line"><span class="number">100</span>:             <span class="comment">// 信号量释放Action</span></div><div class="line"><span class="number">101</span>:             <span class="keyword">final</span> AtomicBoolean semaphoreHasBeenReleased = <span class="keyword">new</span> AtomicBoolean(<span class="keyword">false</span>);</div><div class="line"><span class="number">102</span>:             <span class="keyword">final</span> Action0 singleSemaphoreRelease = <span class="keyword">new</span> Action0() &#123;</div><div class="line"><span class="number">103</span>:                 <span class="meta">@Override</span></div><div class="line"><span class="number">104</span>:                 <span class="function"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title">call</span><span class="params">()</span> </span>&#123;</div><div class="line"><span class="number">105</span>:                     <span class="keyword">if</span> (semaphoreHasBeenReleased.compareAndSet(<span class="keyword">false</span>, <span class="keyword">true</span>)) &#123;</div><div class="line"><span class="number">106</span>:                         fallbackSemaphore.release();</div><div class="line"><span class="number">107</span>:                     &#125;</div><div class="line"><span class="number">108</span>:                 &#125;</div><div class="line"><span class="number">109</span>:             &#125;;</div><div class="line"><span class="number">110</span>: </div><div class="line"><span class="number">111</span>:             Observable&lt;R&gt; fallbackExecutionChain;</div><div class="line"><span class="number">112</span>: </div><div class="line"><span class="number">113</span>:             <span class="comment">// acquire a permit</span></div><div class="line"><span class="number">114</span>:             <span class="keyword">if</span> (fallbackSemaphore.tryAcquire()) &#123;</div><div class="line"><span class="number">115</span>:                 <span class="keyword">try</span> &#123;</div><div class="line"><span class="number">116</span>:                     <span class="keyword">if</span> (isFallbackUserDefined()) &#123;</div><div class="line"><span class="number">117</span>:                         executionHook.onFallbackStart(<span class="keyword">this</span>);</div><div class="line"><span class="number">118</span>:                         fallbackExecutionChain = getFallbackObservable();</div><div class="line"><span class="number">119</span>:                     &#125; <span class="keyword">else</span> &#123;</div><div class="line"><span class="number">120</span>:                         <span class="comment">//same logic as above without the hook invocation</span></div><div class="line"><span class="number">121</span>:                         fallbackExecutionChain = getFallbackObservable();</div><div class="line"><span class="number">122</span>:                     &#125;</div><div class="line"><span class="number">123</span>:                 &#125; <span class="keyword">catch</span> (Throwable ex) &#123;</div><div class="line"><span class="number">124</span>:                     <span class="comment">//If hook or user-fallback throws, then use that as the result of the fallback lookup</span></div><div class="line"><span class="number">125</span>:                     fallbackExecutionChain = Observable.error(ex);</div><div class="line"><span class="number">126</span>:                 &#125;</div><div class="line"><span class="number">127</span>: </div><div class="line"><span class="number">128</span>:                 <span class="comment">// 获得 【回退逻辑 Observable】</span></div><div class="line"><span class="number">129</span>:                 <span class="keyword">return</span> fallbackExecutionChain</div><div class="line"><span class="number">130</span>:                         .doOnEach(setRequestContext)</div><div class="line"><span class="number">131</span>:                         .lift(<span class="keyword">new</span> FallbackHookApplication(_cmd)) <span class="comment">// TODO 【2003】【HOOK】</span></div><div class="line"><span class="number">132</span>:                         .lift(<span class="keyword">new</span> DeprecatedOnFallbackHookApplication(_cmd))</div><div class="line"><span class="number">133</span>:                         .doOnNext(markFallbackEmit)</div><div class="line"><span class="number">134</span>:                         .doOnCompleted(markFallbackCompleted)</div><div class="line"><span class="number">135</span>:                         .onErrorResumeNext(handleFallbackError) <span class="comment">//</span></div><div class="line"><span class="number">136</span>:                         .doOnTerminate(singleSemaphoreRelease)</div><div class="line"><span class="number">137</span>:                         .doOnUnsubscribe(singleSemaphoreRelease);</div><div class="line"><span class="number">138</span>:             &#125; <span class="keyword">else</span> &#123;</div><div class="line"><span class="number">139</span>:                <span class="keyword">return</span> handleFallbackRejectionByEmittingError();</div><div class="line"><span class="number">140</span>:             &#125;</div><div class="line"><span class="number">141</span>:         &#125; <span class="keyword">else</span> &#123;</div><div class="line"><span class="number">142</span>:             <span class="keyword">return</span> handleFallbackDisabledByEmittingError(originalException, failureType, message);</div><div class="line"><span class="number">143</span>:         &#125;</div><div class="line"><span class="number">144</span>:     &#125;</div><div class="line"><span class="number">145</span>: &#125;</div></pre></td></tr></table></figure><ul><li>耐心，这个方法看起来灰常长，也仅限于长，理解成难度很小。</li><li>第 3 行 ：记录 HystrixRequestContext 。</li><li>第 5 至 8 行 ：标记 <code>executionResult</code> 添加( 记录 )事件。</li><li>第 10 至 17 行 ：调用 <code>#isUnrecoverable(Exception)</code> 方法，若异常<strong>不可恢复</strong>，直接返回【异常 Observable】。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1077" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法。</li><li>第 19 至 21 行 ：调用 <code>#isRecoverableError(Exception)</code> 方法，若异常<strong>可恢复</strong>，打印 <code>WARN</code> 日志。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1093" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法。主要针对 <code>java.lang.Error</code> 情况，打印 <code>#isUnrecoverable(Exception)</code> 排除掉的 Error。</li><li>【<em>反向</em>】第 141 至 143 行 ：当配置 <code>HystrixCommandProperties.fallbackEnabled = false</code> ( 默认值 ：<code>true</code> ) ，即失败回退功能<strong>关闭</strong>，调用 <code>#handleFallbackDisabledByEmittingError()</code> ，返回【异常 Observable】。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1047" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法。</li><li>【<em>反向</em>】第 138 至 140 行 ：<strong>失败回退</strong>信号量( TryableSemaphore )【注意，不是<strong>正常执行</strong>信号量】使用失败，调用 <code>#handleFallbackRejectionByEmittingError()</code> ，返回【异常 Observable】。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1038" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法。</li><li>第 23 行 ：当配置 <code>HystrixCommandProperties.fallbackEnabled = true</code> ( 默认值 ：<code>true</code> ) ，即失败回退功能<strong>开启</strong>。</li><li>第 27 至 32 行 ：设置 HystrixRequestContext 的 Action ，使用<strong>第 3 行</strong>记录的 HystrixRequestContext 。</li><li>第 35 至 43 行 ：TODO 【2007】【executionResult】用途</li><li>第 46 至 53 行 ：TODO 【2007】【executionResult】用途</li><li><p>第 56 至 95 行 ：处理回退逻辑执行发生异常的 Func1 ，返回【异常 Observable】。</p><ul><li>第 61 行 ：TODO 【2003】【HOOK】</li><li>第 63 行 ：调用 <code>#getExceptionFromThrowable(Throwable)</code> 方法，获得 Exception 。若 <code>t</code> 的<strong>类型</strong>为 Throwable 时，包装成 Exception 。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1978" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法代码。</li><li>第 68 至 76 行 ：当 <code>fe</code> 的<strong>类型</strong>为 UnsupportedOperationException 时，使用 <code>e</code> + <code>fe</code> 创建 HystrixRuntimeException 。该异常发生于 <code>HystrixCommand#getFallback()</code> <strong>抽象方法</strong>未被覆写。</li><li>第 77 至 86 行 ：当 <code>fe</code> 的<strong>类型</strong>为其他异常时，使用 <code>e</code> + <code>fe</code> 创建 HystrixRuntimeException 。该异常发生于 <code>HystrixCommand#getFallback()</code> 执行发生异常。</li><li>第 89 至 91 行 ：调用 <code>#shouldNotBeWrapped()</code> 方法，判断 <code>originalException</code> 是 ExceptionNotWrappedByHystrix 的<strong>实现</strong>时，即<strong>要求</strong>返回的【异常 Observable】不使用 HystrixRuntimeException 包装。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1057" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法代码。</li><li>第 93 行 ：返回【异常 Observable】，使用 <code>toEmit</code> ( HystrixRuntimeException ) 为异常。</li></ul></li><li><p>第 98 行 ：调用 <code>#getFallbackSemaphore()</code> 方法，获得<strong>失败回退</strong>信号量( TryableSemaphore )对象，点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/AbstractCommand.java#L1241" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法代码。TryableSemaphore 在 <a href="http://www.iocoder.cn/Hystrix/command-execute-first-run/?self">《Hystrix 源码解析 —— 命令执行（一）之正常执行逻辑》「3. TryableSemaphore」</a> 有详细解析。</p></li><li>第 100 至 109 行 ：信号量<strong>释放</strong>的 Action。</li><li>第 114 至 137 行 ：<strong>失败回退</strong>信号量( TryableSemaphore )使用成功，返回【回退逻辑 Observable】。<ul><li>【重要】第 116 至 122 行 ：调用 <code>#getFallbackObservable()</code> 方法，创建【回退逻辑 Observable】。将子类对 <code>HystrixCommand#getFallback()</code> <strong>抽象方法</strong>的执行结果，使用 <code>Observable#just(...)</code> 包装返回。点击 <a href="https://github.com/YunaiV/Hystrix/blob/2655a1323a7b77fc65f31de82b40331797dc018d/hystrix-core/src/main/java/com/netflix/hystrix/HystrixCommand.java#L314" rel="external nofollow noopener noreferrer" target="_blank">链接</a> 查看该方法的代码。<ul><li>第 116 行 ：调用 <code>#isFallbackUserDefined()</code> 方法，返回<strong>命令子类</strong>是否实现 <code>HystrixCommand#getFallback()</code> <strong>抽象方法</strong>。只有已实现( <code>true</code> ) 的情况下，调用 HOOK TODO 【2003】【HOOK】。</li></ul></li><li>第 129 至 137 行 ：获得 【回退逻辑 Observable】。<ul><li>第 131 行 ：// TODO 【2003】【HOOK】</li><li>第 135 行 ：调用 <code>Observable#onErrorResumeNext(...)</code> 方法，实现【失败回退 Observable】执行异常时，返回【异常 Observable】。</li></ul></li></ul></li></ul><hr><p>有两个注意点：</p><ul><li>当命令执行<strong>超时</strong>时，失败回退逻辑使用的是 <strong>HystrixTimer 的线程池</strong>。</li><li>失败回退逻辑，无超时时间，使用要小心。</li></ul><h1 id="666-彩蛋"><a href="#666-彩蛋" class="headerlink" title="666. 彩蛋"></a>666. 彩蛋</h1><p><img src="http://www.iocoder.cn/images/Architecture/2017_12_29/01.png" alt="知识星球"></p><p>比想象中“臭长”的逻辑。</p><p>总的来说，逻辑和 <a href="http://www.iocoder.cn/Hystrix/command-execute-first-run/?self">《Hystrix 源码解析 —— 命令执行（一）之正常执行逻辑》</a> 是很类似的。</p><p>胖友，分享一波朋友圈可好！</p></div><footer class="article-footer clearfix"><div class="article-categories"><span></span> <a class="article-category-link" href="/categories/Hystrix/">Hystrix</a></div><div class="article-share" id="share"><div data-url="http://www.iocoder.cn/Hystrix/command-execute-fourth-fallback/" data-title="Hystrix 源码解析 —— 请求执行（四）之失败回退逻辑 | 芋道源码 —— 纯源码解析博客" data-tsina="null" class="share clearfix"></div></div></footer></article><nav class="article-nav clearfix"><div class="prev"><a href="/Hystrix/command-collapser-execute/" title="Hystrix 源码解析 —— 命令合并执行"><strong>PREVIOUS:</strong><br><span>Hystrix 源码解析 —— 命令合并执行</span></a></div><div class="next"><a href="/Hystrix/command-execute-third-timeout/" title="Hystrix 源码解析 —— 命令执行（三）之执行超时"><strong>NEXT:</strong><br><span>Hystrix 源码解析 —— 命令执行（三）之执行超时</span></a></div></nav></div><div class="openaside"><a class="navbutton" href="#" title="显示侧边栏"></a></div><div id="toc" class="toc-aside"><strong class="toc-title">文章目录</strong><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#1-概述"><span class="toc-number">1.</span> <span class="toc-text">1. 概述</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2-handleFallback"><span class="toc-number">2.</span> <span class="toc-text">2. handleFallback</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#3-handleShortCircuitViaFallback"><span class="toc-number">3.</span> <span class="toc-text">3. #handleShortCircuitViaFallback()</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#4-handleSemaphoreRejectionViaFallback"><span class="toc-number">4.</span> <span class="toc-text">4. #handleSemaphoreRejectionViaFallback()</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#5-handleThreadPoolRejectionViaFallback"><span class="toc-number">5.</span> <span class="toc-text">5. #handleThreadPoolRejectionViaFallback()</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#6-handleTimeoutViaFallback"><span class="toc-number">6.</span> <span class="toc-text">6. #handleTimeoutViaFallback()</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#7-handleFailureViaFallback"><span class="toc-number">7.</span> <span class="toc-text">7. #handleFailureViaFallback()</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#8-getFallbackOrThrowException-…"><span class="toc-number">8.</span> <span class="toc-text">8. #getFallbackOrThrowException(…)</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#666-彩蛋"><span class="toc-number">9.</span> <span class="toc-text">666. 彩蛋</span></a></li></ol></div><div id="asidepart"><aside class="clearfix"><div id="authorInfo"><div><img width="100%" src="/images/common/wechat_mp_2017_07_31_bak.jpg"></div><div class="social-list"></div></div><div class="categorieslist"><p class="asidetitle">微信公众号福利：芋道源码</p><ul><li><a href="/images/common/wechat_mp_2017_07_31_bak.jpg">0. 阅读源码葵花宝典</a></li><li><a href="/images/common/wechat_mp_2017_07_31_bak.jpg">1. RocketMQ / MyCAT / Sharding-JDBC 详细中文注释源码</a></li><li><a href="/images/common/wechat_mp_2017_07_31_bak.jpg">2. 您对于源码的疑问每条留言都将得到认真回复</a></li><li><a href="/images/common/wechat_mp_2017_07_31_bak.jpg">3. 新的源码解析文章实时收到通知，每周六十点更新</a></li><li><a href="/images/common/wechat_mp_2017_07_31_bak.jpg">4. 认真的源码交流微信群</a></li></ul></div><div class="categorieslist"><p class="asidetitle">分类</p><ul><li><a href="/categories/Elastic-Job-Cloud/" title="Elastic-Job-Cloud">Elastic-Job-Cloud<sup>6</sup></a></li><li><a href="/categories/Elastic-Job-Lite/" title="Elastic-Job-Lite">Elastic-Job-Lite<sup>16</sup></a></li><li><a href="/categories/Eureka/" title="Eureka">Eureka<sup>23</sup></a></li><li><a href="/categories/Hystrix/" title="Hystrix">Hystrix<sup>9</sup></a></li><li><a href="/categories/MyCAT/" title="MyCAT">MyCAT<sup>9</sup></a></li><li><a href="/categories/RocketMQ/" title="RocketMQ">RocketMQ<sup>14</sup></a></li><li><a href="/categories/RxJava/" title="RxJava">RxJava<sup>5</sup></a></li><li><a href="/categories/Sharding-JDBC/" title="Sharding-JDBC">Sharding-JDBC<sup>18</sup></a></li><li><a href="/categories/SkyWalking/" title="SkyWalking">SkyWalking<sup>35</sup></a></li><li><a href="/categories/Spring-Cloud-Gateway/" title="Spring-Cloud-Gateway">Spring-Cloud-Gateway<sup>24</sup></a></li><li><a href="/categories/TCC-Transaction/" title="TCC-Transaction">TCC-Transaction<sup>7</sup></a></li><li><a href="/categories/TiKV/" title="TiKV">TiKV<sup>2</sup></a></li><li><a href="/categories/工作内推/" title="工作内推">工作内推<sup>6</sup></a></li><li><a href="/categories/芋道源码的周八/" title="芋道源码的周八">芋道源码的周八<sup>10</sup></a></li></ul></div><div id="authorInfo2"><div><img width="100%" src="/images/common/zsxq/02.png"></div></div></aside></div></div><footer><div id="footer"><img src="http://www.iocoder.cn/images/common/wechat_mp_simple.png" style="display:none"><p class="copyright">© 2018 <a href="http://www.iocoder.cn" target="_blank" title="芋道源码">芋道源码</a> && <span style="display:inline" id="busuanzi_container_site_uv">总访客数 <span id="busuanzi_value_site_uv" font="微软雅黑"></span> 次</span> && <span style="display:inline" id="busuanzi_container_site_pv">总访问量 <span id="busuanzi_value_site_pv" font="微软雅黑"></span> 次</span> && Hosted by <a href="https://pages.coding.me" style="font-weight:700" rel="external nofollow noopener noreferrer" target="_blank">Coding Pages</a></p></div><div class="copyright">沪ICP备17037075号-1</div></footer><script src="/js/jquery-2.1.0.min.js"></script><script type="text/javascript">$(document).ready(function(){function e(){"number"==typeof window.innerWidth?n=window.innerWidth:document.documentElement&&document.documentElement.clientWidth&&(n=document.documentElement.clientWidth)}$(".navbar").click(function(){$("header nav").toggleClass("shownav")});var n=0,s=$("#main"),a=$("#asidepart"),o=$(".closeaside"),d=$(".openaside");$(window).resize(function(){e(),n>=1024?$("header nav").removeClass("shownav"):(s.removeClass("moveMain"),a.css("display","block").removeClass("fadeOut"),d.css("display","none"),$("#toc.toc-aside").css("display","none"))}),o.click(function(){a.addClass("fadeOut").css("display","none"),d.css("display","block").addClass("fadeIn"),s.addClass("moveMain")}),d.click(function(){d.css("display","none").removeClass("beforeFadeIn"),a.css("display","block").removeClass("fadeOut").addClass("fadeIn"),s.removeClass("moveMain")}),$(window).scroll(function(){d.css("top",Math.max(80,260-$(this).scrollTop()))})})</script><script type="text/javascript">$(document).ready(function(){var a=$(".article-content>iframe"),t=$(".article-content>embed"),n=$("#toc");$("article h2");ah=$("article h2"),ta=$("#toc.toc-aside"),o=$(".openaside"),c=$(".closeaside"),a.length>0&&a.wrap('<div class="video-container" />'),t.length>0&&t.wrap('<div class="video-container" />'),0==ah.length?n.css("display","none"):(c.click(function(){ta.css("display","block").addClass("fadeIn")}),o.click(function(){ta.css("display","none")}),$(window).scroll(function(){ta.css("top",Math.max(140,320-$(this).scrollTop()))}))})</script><script type="text/javascript">$(document).ready(function(){var t=$(".share"),a=t.attr("data-url"),e=encodeURIComponent(a),r=['<a href="#" class="overlay" id="qrcode"></a>','<div class="qrcode clearfix"><span>扫描二维码分享到微信朋友圈</span><a class="qrclose" href="#share"></a><strong>Loading...Please wait</strong><img id="qrcode-pic" data-src="http://s.jiathis.com/qrcode.php?url='+e+'"/></div>','<a href="#textlogo" class="article-back-to-top" title="Top"></a>','<a href="https://www.facebook.com/sharer.php?u='+e+'" class="article-share-facebook" target="_blank" title="Facebook"></a>','<a href="#qrcode" class="article-share-qrcode" title="QRcode"></a>','<a href="https://twitter.com/intent/tweet?url='+e+'" class="article-share-twitter" target="_blank" title="Twitter"></a>','<a href="http://service.weibo.com/share/share.php?title='+t.attr("data-title")+"&url="+e+"&ralateUid="+t.attr("data-tsina")+'&searchPic=true&style=number" class="article-share-weibo" target="_blank" title="Weibo"></a>','<span title="Share to"></span>'].join("");t.append(r),$(".article-share-qrcode").click(function(){var t=$("#qrcode-pic").attr("data-src");$("#qrcode-pic").attr("src",t),$("#qrcode-pic").load(function(){$(".qrcode strong").text(" ")})})})</script><link rel="stylesheet" href="/alert/css/alert.css"><script src="/alert/js/alert.js"></script><script src="/js/jquery.cookie.js"></script><script src="/js/util.js"></script><script type="text/javascript">!function(e,a,t,n,c,o,s){e.GoogleAnalyticsObject=c,e[c]=e[c]||function(){(e[c].q=e[c].q||[]).push(arguments)},e[c].l=1*new Date,o=a.createElement(t),s=a.getElementsByTagName(t)[0],o.async=1,o.src="//www.google-analytics.com/analytics.js",s.parentNode.insertBefore(o,s)}(window,document,"script",0,"ga"),ga("create","UA-107572620-1","auto"),ga("send","pageview")</script></body>